b2sdk/scan/folder.py

```diff
@@ -204,39 +212,52 @@ def _walk_relative_paths(
         #
         # This is because in Unicode '.' comes before '/', which comes before '0'.
         names = []  # list of (name, local_path, relative_file_path)
-        for name in os.listdir(local_dir):
-            # We expect listdir() to return unicode if dir_path is unicode.
-            # If the file name is not valid, based on the file system
-            # encoding, then listdir() will return un-decoded str/bytes.
-            if not isinstance(name, str):
-                name = self._handle_non_unicode_file_name(name)
+
+        visited_symlinks = visited_symlinks or set()
+
+        if local_dir.is_symlink():
+            real_path = local_dir.resolve()
+            inode_number = real_path.stat().st_ino
+            if inode_number in visited_symlinks:
+                # Infinite symlink loop detected, report warning and skip symlink
+                if reporter is not None:
+                    reporter.circular_symlink_skipped(str(local_dir))
+                return
+
+            visited_symlinks.add(inode_number)
```

Comment:
the first and last line of that range is an operation that needs to find out whether the set contains an item, the second actually adds it. I think for big sets it would be significantly faster to do `len(visited_symlinks)`, `visited_symlinks.add(inode_number)` and `len()` again to see if we succeeded in adding an element (because if we didn't then it was already there). A  great opportunity for you to use the `timeit` module!

Reply:
Nice one!

```
Time elapsed for original code: 4.718000127468258e-06
Time elapsed for optimized code: 2.958000550279394e-06
```

Reply:
🤔 is the optimized code faster for an empty set too?

Reply:
No, for smaller sets the oiriginal code is better or similar:

Empty set:
```
Time elapsed for original code: 1.319000034527562e-06
Time elapsed for optimized code: 2.3310000187848345e-06
```

1000 elements:
```
Time elapsed for original code: 1.0374999988016498e-05
Time elapsed for optimized code: 1.1050000011891825e-05
```

1m:
```
Time elapsed for original code: 5.107999982101319e-06
Time elapsed for optimized code: 2.774000108729524e-06
```

10m:
```
Time elapsed for original code: 6.407000000763219e-06
Time elapsed for optimized code: 3.318000040053448e-06
```